### DISCUSSION TITLE
MONAI dataloader producing samples of inconsistent dimension

### DISCUSSION THREAD

**Discussion by philnovv** [Project-MONAI/MONAI#7731] [General]
Hi all, here is my set of transforms I'm using:

 ```
[
            LoadImaged(
                keys=["image", "label"],
                allow_missing_keys=True),
            EnsureChannelFirstd(
                keys=["image", "label"],
                allow_missing_keys=True),
            Orientationd(
                keys=["image", "label"],
                axcodes="RAS"),
            Spacingd(
                keys=["image", "label"],
                pixdim=config["voxel_size"],
                mode=("bilinear", "nearest")),
            NormalizeIntensityd(
                keys=["image"],
                channel_wise=True),
            CastToTyped(
                keys=["label"],
                dtype=(torch.int16)),
            SpatialPadd(
                keys=["image", "label"],
                spatial_size=config["train_patch_size_voxels"]),
            RandCropByPosNegLabeld(
                keys=["image", "label"],
                label_key="label",
                spatial_size=config["train_patch_size_voxels"],
                pos=config['pos_sample_ratio'],
                neg=1,
                allow_smaller=False)
        ])
```
Using this set of transforms, I sometimes obtain image samples with different dimensions, e.g.
[1,92,92,92] and [1,92,92,87].

My question is: how is this possible? My understanding is that if the image size is less than e.g. 92 in any dimension, then SpatialPadd will pad it to 92 where necessary. Since this happens before random crop, shouldn't the image always have dimensions greater than or equal to the crop dimensions? 

A second issue: despite calling "allow_smaller=False" in the random cropping init, no error is raised my MONAI when the inconsistent samples are produced.

Any thoughts would be greatly appreciated!

**Comment by KumoLiu** [comment]
Hi @philnovv, you should move the `SpatialPadd` after `RandCropByPosNegLabeld` or add another `SpatialPadd`.

> how is this possible? My understanding is that if the image size is less than e.g. 92 in any dimension, then SpatialPadd will pad it to 92 where necessary. Since this happens before random crop, shouldn't the image always have dimensions greater than or equal to the crop dimensions?

Yes, `SpatialPadd` will ensure all your output larger or equal to 92. But `RandCropByPosNegLabeld` will random generate the crop center which can not ensure your output still have 92.

> A second issue: despite calling "allow_smaller=False" in the random cropping init, no error is raised my MONAI when the inconsistent samples are produced.

`allow_smaller` is used to check whether your data is smaller than the roi size before crop since you have `SpatialPadd`. So after it, all the image size will >= 92 then will not raise exception.
https://github.com/Project-MONAI/MONAI/blob/e1a69b03c86ce065db2816b696ea4a6b57d46435/monai/transforms/croppad/dictionary.py#L919

Hope it helps, thanks.
